res/res_format_attr_red.c

```diff
@@ -0,0 +1,283 @@
+/*
+ * Asterisk -- An open source telephony toolkit.
+ *
+ * Copyright (C) 2025, GILAWA Ltd
+ *
+ * Henning Westerholt <hw@gilawa.com>
+ *
+ * See http://www.asterisk.org for more information about
+ * the Asterisk project. Please do not directly contact
+ * any of the maintainers of this project for assistance;
+ * the project provides a web site, mailing lists and IRC
+ * channels for your use.
+ *
+ * This program is free software, distributed under the terms of
+ * the GNU General Public License Version 2. See the LICENSE file
+ * at the top of the source tree.
+ */
+
+/*!
+ * \file
+ * \brief RED format attribute interface
+ *
+ * \author Henning Westerholt <hw@gilawa.com>
+ *
+ * \note https://www.rfc-editor.org/rfc/rfc4103.html
+ */
+
+/*** MODULEINFO
+	<support_level>core</support_level>
+ ***/
+
+#include "asterisk.h"
+
+#include <ctype.h>                      /* for tolower */
+
+#include "asterisk/module.h"
+#include "asterisk/format.h"
+#include "asterisk/logger.h"            /* for ast_log, LOG_WARNING */
+#include "asterisk/strings.h"           /* for ast_str_append */
+#include "asterisk/utils.h"             /* for ast_free */
+#include "asterisk/rtp_engine.h"        /* for AST_RED_MAX_GENERATION */
+
+/*
+ * From RFC 4103: "Therefore, text/t140 is RECOMMENDED
+ * to be the only payload type in the RTP stream."
+ * For this reason we only support uniform payload types.
+ *
+ * If in the future this attribute should be used also for other redundant rtp streams,
+ * it needs to be adapted together with the other infrastructure supporting it.
+ */
+struct red_attr {
+	int red_num_gen; /* Number of generations */
+	int red_payload; /* Payload for each generation (actually related to t140) */
+	int cps;         /* Characters per second */
+};
+
+static struct red_attr default_red_attr = {
+	.red_num_gen = 2,
+	.red_payload = 98, /* default type that two test clients used */
+	.cps = 30
+};
+
+static void red_destroy(struct ast_format *format)
+{
+	struct red_attr *attr = ast_format_get_attribute_data(format);
+	ast_free(attr);
+}
+
+static int red_clone(const struct ast_format *src, struct ast_format *dst)
+{
+	struct red_attr *original = ast_format_get_attribute_data(src);
+	struct red_attr *attr = ast_malloc(sizeof(*attr));
+
+	if (!attr) {
+		return -1;
+	}
+
+	if (original) {
+		*attr = *original;
+	} else {
+		*attr = default_red_attr;
+	}
+
+	ast_format_set_attribute_data(dst, attr);
+	return 0;
+}
+
+static struct ast_format *red_parse_sdp_fmtp(const struct ast_format *format, const char *attributes)
+{
+	char *attribs = ast_strdupa(attributes);
+	char red_cp_data[30];
+	char *red_cp = red_cp_data;
+	char *rest = NULL;
+	struct ast_format *cloned;
+	struct red_attr *attr;
+	int red_num_gen = -1;
+	int red_payload = -1;
+
+	cloned = ast_format_clone(format);
+	if (!cloned) {
+		return NULL;
+	}
+	attr = ast_format_get_attribute_data(cloned);
+
+	strcpy(red_cp, attribs);
```

Comment:
If `strlen(attribs) > 29` we're going to have a bad time. Maybe use `ast_strdupa(attribs)` to allocate/initialize `red_cp` instead?

Reply:
good suggestion